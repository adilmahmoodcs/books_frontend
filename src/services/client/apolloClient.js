import { ApolloClient, InMemoryCache, HttpLink } from '@apollo/client'
import fetch from 'cross-fetch'

export const client = new ApolloClient({
  link: new HttpLink({
    uri: `${import.meta.env.VITE_API_URL}/graphql`,
    fetch
  }),
  cache: new InMemoryCache()
})
