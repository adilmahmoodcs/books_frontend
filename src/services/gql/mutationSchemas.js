import { gql } from "@apollo/client";

export const CREATE_AUTHOR = gql`
  mutation CreateAuthor($author: AuthorCreateInput!) {
    authorCreate(input: $author) {
      author {
        id
        firstname
        surname
        dob
      }
    }
  }
`;

export const DELETE_AUTHOR = gql`
  mutation DeleteAuthor($authorId: AuthorDeleteInput!) {
    authorDelete(input: $authorId) {
      author {
        id
      }
    }
  }
`;

export const CREATE_BOOK = gql`
  mutation CreateBook($book: BookCreateInput!) {
    bookCreate(input: $book) {
      book {
        id
        title
      }
    }
  }
`;

export const DELETE_BOOK = gql`
  mutation DeleteBook($bookId: BookDeleteInput!) {
    bookDelete(input: $bookId) {
      book {
        id
      }
    }
  }
`;
