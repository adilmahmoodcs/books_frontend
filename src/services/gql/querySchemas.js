import { gql } from "@apollo/client";

export const GET_AUTHORS = gql`
  query GetAuthors {
    authors {
      id
      firstname
      surname
      dob
    }
  }
`;

export const GET_AUTHOR_BY_ID = (authorId) => gql`
  query GetAuthorById {
    author(id: ${authorId}) {
      firstname
      surname
      dob
      books {
        id
        title
        genre
        publishYear
      }
    }
  }
`;
