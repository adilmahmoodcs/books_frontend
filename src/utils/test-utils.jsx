/* eslint-disable import/export */
import { ApolloProvider } from "@apollo/client";
import { render } from "@testing-library/react";
import { BrowserRouter } from "react-router-dom";
import client from "../services/client/apolloClient";

const customRender = (ui, options = {}) =>
  render(ui, {
    // wrap provider(s) here if needed
    wrapper: ({ children }) => (
      // <ApolloProvider client={client}>
        <BrowserRouter>{children}</BrowserRouter>
      // </ApolloProvider>
    ),

    ...options,
  });

export * from "@testing-library/react";
export { default as userEvent } from "@testing-library/user-event";

// override render export
export { customRender as render };
