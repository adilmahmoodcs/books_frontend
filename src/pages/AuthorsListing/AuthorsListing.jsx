import { useQuery } from "@apollo/client";
import { Box, Button, Typography } from "@mui/material";
import React, { useState } from "react";
import { AddAuthor } from "../../components";
import { GET_AUTHORS } from "../../services/gql/querySchemas";
import { Table } from "./Components";

function AuthorsListing(props) {
  const { data, error, loading, refetch } = useQuery(GET_AUTHORS);

  const [open, setOpen] = useState(false);

  if (loading) return <p>Loading...</p>;

  if (error) return <p>Error...</p>;

  return (
    <Box sx={{}}>
      <Box
        sx={{
          display: "flex",
          justifyContent: "space-between",
          alignItems: "center",
          mb: 2,
        }}
      >
        <Typography variant="h5" sx={{ mb: 2 }}>
          Authors
        </Typography>

        <Button variant="contained" onClick={() => setOpen(true)}>
          Add New Author
        </Button>
      </Box>

      <Table data={data.authors} />

      <AddAuthor open={open} handleClose={() => setOpen(false)} />
    </Box>
  );
}

AuthorsListing.propTypes = {};

export default AuthorsListing;
