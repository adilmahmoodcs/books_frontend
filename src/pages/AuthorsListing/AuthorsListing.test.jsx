import { MockedProvider } from "@apollo/client/testing";
import { describe, expect, it } from "vitest";
import { GET_AUTHORS } from "../../services/gql/querySchemas";
import { render, waitFor } from "../../utils/test-utils";
import AuthorsListing from "./AuthorsListing";

const mocks = [
  {
    request: {
      query: GET_AUTHORS,
    },
    result: {
      data: {
        authors: [
          {
            dob: "2022-03-09T19:00:00Z",
            firstname: "Jane",
            id: "1",
            surname: "Smith",
            __typename: "Author",
          },
        ],
      },
    },
  },
];

const WrappedComponent = () => (
  <MockedProvider mocks={mocks}>
    <AuthorsListing />
  </MockedProvider>
);

describe("it tests Author listing", () => {
  it("query is fetched", async () => {
    const { container } = render(<WrappedComponent />);
    await waitFor(() => new Promise((res) => setTimeout(res, 0)));
    expect(container).toMatchSnapshot();
  });
});
