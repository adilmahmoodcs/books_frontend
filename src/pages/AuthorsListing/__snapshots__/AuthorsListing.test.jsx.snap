// Vitest Snapshot v1

exports[`it tests Author listing > query is fetched 1`] = `
<div>
  <div
    class="MuiBox-root css-0"
  >
    <div
      class="MuiBox-root css-1kw3y0a"
    >
      <h5
        class="MuiTypography-root MuiTypography-h5 css-ch2k2f-MuiTypography-root"
      >
        Authors
      </h5>
      <button
        class="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root  css-sghohy-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="button"
      >
        Add New Author
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiTableContainer-root css-11xur9t-MuiPaper-root-MuiTableContainer-root"
    >
      <table
        aria-label="simple table"
        class="MuiTable-root css-quj9j6-MuiTable-root"
      >
        <thead
          class="MuiTableHead-root css-15wwp11-MuiTableHead-root"
        >
          <tr
            class="MuiTableRow-root MuiTableRow-head css-1q1u3t4-MuiTableRow-root"
          >
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignCenter MuiTableCell-sizeMedium css-1ndpvdd-MuiTableCell-root"
              scope="col"
            >
              Author Id
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignCenter MuiTableCell-sizeMedium css-1ndpvdd-MuiTableCell-root"
              scope="col"
            >
              First Name
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignCenter MuiTableCell-sizeMedium css-1ndpvdd-MuiTableCell-root"
              scope="col"
            >
              Surname
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignCenter MuiTableCell-sizeMedium css-1ndpvdd-MuiTableCell-root"
              scope="col"
            >
              Date of Birth
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignCenter MuiTableCell-sizeMedium css-1ndpvdd-MuiTableCell-root"
              scope="col"
            >
              Books
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignCenter MuiTableCell-sizeMedium css-1ndpvdd-MuiTableCell-root"
              scope="col"
            />
          </tr>
        </thead>
        <tbody
          class="MuiTableBody-root css-apqrd9-MuiTableBody-root"
        >
          <tr
            class="MuiTableRow-root css-7esopy-MuiTableRow-root"
          >
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
            >
              1
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
            >
              Jane
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
            >
              Smith
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
            >
              Thu Mar 10 2022
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
            >
              <a
                class="MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButtonBase-root  css-1rwt2y5-MuiButtonBase-root-MuiButton-root"
                href="/1/books"
                tabindex="0"
              >
                Books
                <span
                  class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                />
              </a>
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight MuiTableCell-sizeMedium css-177gid-MuiTableCell-root"
            >
              <button
                class="MuiLoadingButton-root MuiButton-root MuiButton-outlined MuiButton-outlinedError MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButtonBase-root css-1mlxqnc-MuiButtonBase-root-MuiButton-root-MuiLoadingButton-root"
                id="mui-1"
                tabindex="0"
                type="button"
              >
                <span
                  class="MuiButton-startIcon MuiButton-iconSizeMedium css-1d6wzja-MuiButton-startIcon"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                    data-testid="DeleteForeverIcon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zm2.46-7.12 1.41-1.41L12 12.59l2.12-2.12 1.41 1.41L13.41 14l2.12 2.12-1.41 1.41L12 15.41l-2.12 2.12-1.41-1.41L10.59 14l-2.13-2.12zM15.5 4l-1-1h-5l-1 1H5v2h14V4z"
                    />
                  </svg>
                </span>
                Delete
                <span
                  class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                />
              </button>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</div>
`;
